boa/src/builtins/math/mod.rs

```diff
@@ -502,11 +615,29 @@ impl Math {
     /// [spec]: https://tc39.es/ecma262/#sec-math.max
     /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/max
     pub(crate) fn max(_: &JsValue, args: &[JsValue], context: &mut Context) -> JsResult<JsValue> {
+        // 1. Let coerced be a new empty List.
+        // 2. For each element arg of args, do
+        // b. Append n to coerced.
+        // 3. Let highest be -∞𝔽.
         let mut max = f64::NEG_INFINITY;
+
+        // 4. For each element number of coerced, do
         for arg in args {
+            // a. Let n be ? ToNumber(arg).
             let num = arg.to_number(context)?;
-            max = max.max(num);
+
+            max = if max.is_nan() {
+                continue;
+            } else if num.is_nan() {
+                // a. If number is NaN, return NaN.
+                f64::NAN
```

Comment:
In these cases the loop should be short-circuiting (returning immediately) and you are not doing it, please change these both to `return Ok(f64::NAN);`

Reply:
I thought this too, but it is expected that `ToNumber` is called on every `arg`. There is actually a test that in the 262 suite for that behavior.

Reply:
Yeah, I tried short circuiting but some tests failed. The reason for that is to prioritize conversion errors over a NaN result.

Reply:
Oh, because we are doing it all in 1 loop rather than 2. I guess it's hard to know whether it is worth it though.

Reply:
I was trying to preserve the original implementation, but I could reimplement it using a vector, and it shouldn't make too much of a difference aside from the allocation, with the advantage of being a lot more readable. What do you think?

Reply:
I'm okay with leaving it like this, I think for most use cases this is probably the best choice.